accepted/2024/[TS]PMUI-Readme-rendering.md

```diff
@@ -0,0 +1,207 @@
+            "@type": "RegistrationsBaseUrl/6.12.0"
+        },
+        ...
+    ]
+}
+```
+###### Example Response
+```json
+{
+    "@id": "https://apidev.nugettest.org/v3/registration5-gz-semver2/newtonsoft.json/index.json",
+    "items": [
+        {
+            ...,
+            "items": [ 
+                ...,
+                {
+                    "@id": "https://apidev.nugettest.org/v3/registration5-gz-semver2/newtonsoft.json/13.0.3.json",
+                    "@type": "Package",
+                    "catalogEntry": {
+                        "@id": "https://apidev.nugettest.org/v3/catalog0/data/2024.08.19.16.55.59/newtonsoft.json.13.0.3.json",
+                        "@type": "PackageDetails",
+                        "authors": "James Newton-King",
+                        "dependencyGroups": [ ... ],
+                        "description": "Json.NET is a popular high-performance JSON framework for .NET",
+                        "iconUrl": "https://apidev.nugettest.org/v3-flatcontainer/newtonsoft.json/13.0.3/icon",
+                        "id": "Newtonsoft.Json",
+                        "language": "",
+                        "licenseExpression": "MIT",
+                        "licenseUrl": "https://dev.nugettest.org/packages/Newtonsoft.Json/13.0.3/license",
+                        "readmeUrl": "https://dev.nugettest.org/packages/Newtonsoft.Json/13.0.3#show-readme-container",
+                        //New field
+                        "rawReadmeUrl": "https://apidev.nugettest.org/v3/flatcontainer/newtonsoft.json/13.0.3/readme",
+                        "listed": true,
+                        "minClientVersion": "2.12",
+                        "packageContent": "https://apidev.nugettest.org/v3-flatcontainer/newtonsoft.json/13.0.3/newtonsoft.json.13.0.3.nupkg",
+                        "projectUrl": "https://www.newtonsoft.com/json",
+                        "published": "2023-04-25T14:48:53.817+00:00",
+                        "requireLicenseAcceptance": false,
+                        "summary": "",
+                        "tags": [
+                            "json"
+                        ],
+                        "title": "Json.NET",
+                        "version": "13.0.3"
+                    },
+                    "packageContent": "https://apidev.nugettest.org/v3-flatcontainer/newtonsoft.json/13.0.3/newtonsoft.json.13.0.3.nupkg",
+                    "registration": "https://apidev.nugettest.org/v3/registration5-gz-semver2/newtonsoft.json/index.json"
+                },
+                ...
+            ],
+            ...
+        }
+    ],
+    ...
+}
+```
+##### PackageBaseAddress/6.12.0
+
+A new version of the [package content](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource) resource type which will include a url definition for downloading the README.
+
+```json
+{
+    "version": "3.0.0",
+    "resources": [
+        ...,
+        {
```

Comment:
Why isn't this a template like `ReportAbuseUriTemplate`?
it seems to fit that better. 

Example:
```json
        {
            "@id": "https://www.nuget.org/packages/{id}/{version}/ReportAbuse",
            "@type": "ReportAbuseUriTemplate/3.0.0-rc",
            "comment": "URI template used by NuGet Client to construct Report Abuse URL for packages used by RC clients"
        }
```

Reply:
The [package content](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource) resource describes itself an endpoint for fetching the package's contents. So, I think it makes sense to include it there. While I don't think we implement it in the code this endpoint can be used to download the nuspec and the nupkg.

Reply:
This is a server feedback, but probably a good idea to update the resource comment to indicate that readme can be downloaded as such as well. 

Today, only the nupkg is suggested within the index.json. 

cc @joelverhagen

Reply:
The comment is in NuGet.Services.Index. Whoever updates the service index with this protocol can update the comment.
I updated https://github.com/NuGet/NuGetGallery/issues/10116 with the suggestion.

Reply:
@zivkan brought up a good point about this. Is the expectation that newer versions of an API are backwards compatible with all versions?

PackageBaseAddress/6.12.0 (We can download README)
PackageBaseAddress/6.15.0 (We can download Icons)

From the client perspective should we expect the endpoint provided by PackageBaseAddress/6.15.0 to also allow downloading the README?

Reply:
pedantic: icon isn't a good example, because both search results and registration blobs both contain icon URL. So, there's no need for client to ever probe a generated endpoint based on resource version. Exactly how I wish raw readme URL was implement  😁 

But yes, the point is that if there's some feature in the future, will this design force servers to opt into readme probing if they want to implement the new feature?

I don't know enough about how `SourceRepository.GetResourceAsync` works

Reply:
I don't know if it's specified explicitly anywhere, but everything we have done so far, resources are only additive.

Reply:
Only additive meaning that if we got back a service index with only the `PackageBaseAddress/6.15.0` resource type we would expect it to comply with all the previous versions as well? @joelverhagen  is that how NuGet.org has been operating as well?

Reply:
Yes, I think we'd expect it to comply to previous versions. 

Doesn't mean we can't change it, but if we did, we'd need to better define the protocol and live with the chances that there might be bugs. 

Another thing to consider is that even if we did try to load a readme, we'd fail and it'd be something we'd handle, since readmes aren't mandatory, so in practice it might not even matter that  much.

Reply:
Yea but servers may not want to be exposing themselves to additional authenticated calls to an endpoint that they don't support. I think implementing something similar to the ReportAbuseTemplate would actually make more sense given the implications. 

```json
        {
            "@id": "https://apidev.nugettest.org/v3/flatcontainer/{LOWER_ID}/{LOWER_VERSION}/readme",
            "@type": "ReadmeUriTemplate/16.13",
            "comment": "URI template used by NuGet Client to construct a URL for downloading a package's README"
        }
```

Potentially expanding the allowable variables to be:
- {LOWER_ID} 
- {LOWER_VERSION}
- {UPPER_ID}
- {UPPER_VERSION}
- {ID} Case doesn't matter
- {VERSION} Case doesn't matter

@joelverhagen do you have thoughts on this from a NuGet.org perspective?

Reply:
We should only support lowercase id/version, the same way flat container down. 
We normalize to lowercase everywhere in product, for example the global packages folder is fully lowercase. 
We don't really want to add ambiguity in the protocol.

Reply:
**tldr;** `ReadmeUriTemplate` seems better at least from a tactical perspective, great idea @nkolev92!

Readme URI template seems just fine too. Regarding the conversation about whether resources are addition, when to do breaking changes, etc:

There is no rule that says resource `Foo/Y` has all of the capabilities of `Foo/X` and more (given Y > X). Remember, resource types are opaque strings and the version part of the resource type string is for display purposes. No version parsing is done. In fact any such rule would be problematic for when we DO want to make breaking changes. 

Expanding this concept out a bit. In a parallel universe we could call have called `PackageBaseAddress/3.0.0` as `PackageBaseAddressWithNuspec` and then `PackageBaseAddress/6.12.0` as `PackageBaseAddressWithReadme`. Then later when we add icons we could call it `PackageBaseAddressWithIcons`. (setting aside the valid point that @zivkan made about icons already being solved -- let's use it for illustrative purposes)

Generally speaking, each resource version should describe what it does and what relationship it has with the other versions. It's certainly most simple to always be additive, but it's not required. Take an example `RegistrationsBaseUrl/3.4.0`: this had a breaking change of introducing "gzipped responses". Does it have all of the capabilities of the previous (lower) version `RegistrationsBaseUrl/3.0.0-rc`? No, the capability of returning responses without transfer encoding was lost (non-additive). Note that `RegistrationsBaseUrl/3.6.0` explicitly states that it also has `gzip` responses: https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource#registrationsbaseurl360. Certainly docs could be improved to explicitly state SemVer 1.0.0 versions are included also but docs are easy to improve over time.

If you don't like the nuance of transfer encoding and additive/breaking, then yes you could say all changes have been additive. But that's because we haven't removed scenarios from V3 before. And there is nothing in the docs or protocol that says it has to be this way.

Would `PackageBaseAddressWithIcons` imply there are readmes? Only if we want it to and document it as such. If we didn't want it to do that, we would say NuGet.org needs to keep `PackageBaseAddressWithReadme` in the service index when it adds `PackageBaseAddressWithIcons` to declare its continued support of readmes. But service indexes need to do this anyways! Because older versions are looking for `PackageBaseAddressWithReadme` and not `PackageBaseAddressWithIcons` and they should continue to work. For newer clients that know about icons and readmes, they would look for `PackageBaseAddressWithReadme` to conditionally fetch READMEs and `PackageBaseAddressWithIcons` to conditionally fetch icons. 

So when we want to release a new `PackageBaseAddress` after README, we would just say `PackageBaseAddress/6.15.0` does not declare support for READMEs. We could even say it is a breaking change but that perhaps too strong of language because `PackageBaseAddress/6.15.0` and `PackageBaseAddress/6.12.0` and co-exist with the same URL.

Anyways, that's the theory/design around service index.

The `ReadmeUriTemplate` resource seems great. I don't feel strongly one way or another from a protocol perspective. But from a tactical perspective it has a very good benefit: this could even provide China customers immediate support for readmes via the `globalcdn` DNS name (or a new one we make to allow Microsoft CDN to remain as primary), e.g.
`https://globalcdn.nuget.org/v3-flatcontainer/{LOWER_ID}/{LOWER_VERSION}/readme`. Perhaps an okay short-term solution until we get readmes replicated into China storage.

Reply:
I like it, updated.